contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,582 @@
+        internal
+        returns (uint256 totalOutputTokenBought)
+    {
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredRedemptionComponents(_issuanceModule, _setToken, _amountSetToken);
+        for (uint256 i = 0; i < _swaps.length; i++) {
+            require(components[i] == address(_swaps[i].sellToken), "ExchangeIssuance: COMPONENT / QUOTE ADDRESS MISMATCH");
+            require(address(_swaps[i].buyToken) == address(_outputToken), "ExchangeIssuance: INVALID BUY TOKEN");
+            uint256 maxAmountSell = componentUnits[i];
+
+            require(maxAmountSell <= IERC20(components[i]).balanceOf(address(this)), "ExchangeIssuance: INSUFFICIENT COMPONENT BALANCE");
+
+            uint256 outputTokenAmountBought;
+            uint256 componentAmountSold;
+
+            // If the component is equal to the input token we don't have to trade
+            if(components[i] == address(_swaps[i].buyToken)) {
+                outputTokenAmountBought = maxAmountSell;
+                componentAmountSold = maxAmountSell;
+            }
+            else {
+                _safeApprove(_swaps[i].sellToken, address(swapTarget), maxAmountSell);
+                (outputTokenAmountBought, componentAmountSold) = _fillQuote(_swaps[i]);
+            }
+
+            require(maxAmountSell >= componentAmountSold, "ExchangeIssuance: OVERSOLD COMPONENT");
+            totalOutputTokenBought = totalOutputTokenBought.add(outputTokenAmountBought);
+        }
+    }
+
+    /**
+     * Execute a 0x Swap quote
+     *
+     * @param _quote          Swap quote as returned by 0x API
+     *
+     * @return boughtAmount   The amount of _quote.buyToken obtained
+     * @return spentAmount    The amount of _quote.sellToken spent
+     */
+    function _fillQuote(
+        ZeroExSwapQuote memory _quote
+    )
+        internal
+        returns(uint256 boughtAmount, uint256 spentAmount)
+    {
+        uint256 buyTokenBalanceBefore = _quote.buyToken.balanceOf(address(this));
+        uint256 sellTokenBalanceBefore = _quote.sellToken.balanceOf(address(this));
+
+
+        (bool success, bytes memory returndata) = swapTarget.call(_quote.swapCallData);
+
+        // Forwarding errors including new custom errors
+        // Taken from: https://ethereum.stackexchange.com/a/111187/73805
+        if (!success) {
+            if (returndata.length == 0) revert();
+            assembly {
+                revert(add(32, returndata), mload(returndata))
+            }
+        }
+
+        boughtAmount = _quote.buyToken.balanceOf(address(this)).sub(buyTokenBalanceBefore);
+        spentAmount = sellTokenBalanceBefore.sub(_quote.sellToken.balanceOf(address(this)));
+        emit BoughtTokens(_quote.sellToken, _quote.buyToken, boughtAmount);
+    }
+
+    /**
+     * Redeems a given amount of SetToken.
```

Comment:
Can we add a note to this comment to explain that it also transfers the token from the caller?

Reply:
Done. Also added a note that the components that the set is redeemed for will be received on this contracts address.